### PR TITLE
feat: enable searching for repositories with Github adapter

### PR DESCRIPTION
**Pull request by jugaltheshah**
This PR adds functionality to enable additional Github search types. While currently limited to Github code and repository type searches, the structure makes it easy to add additional search types in the future. Repository search looks at repo metadata such as name, description, size, follows, forks, topics, etc, and finds repos matching the criteria provided.

### CODE REVIEW

File: src/adapters/github.ts
Comment (nwalters512) [inline src/adapters/github.ts:65]: Nit: we don't really refer to these as parameters elsewhere - let's drop that from the error message.

```suggestion
      throw new Error(`"search_type" must be one of the following: ${Array.from(GithubAdapter.validSearchTypes).map(e => `'${e}'`).join(', ')}`);
```

File: package.json
Comment (aorinevo) [inline package.json:49]: Outside the scope of this PR, but didn't the hapi modules get descoped in transitioning the project over to new maintainers?  Any updates on this?

File: src/adapters/github.ts
Comment (nwalters512) [inline src/adapters/github.ts:22]: Micro-nit: Might be more conventional to declare this outside the scope of the class instead of declaring it as a static member, i.e.

```ts
const VALID_SEARCH_TYPES = new Set(['code', 'repositories']);

class GithubAdapter extends GitAdapter {
  // ...
}
```

Should make it slightly clearer that this is a constant. Better yet, we can include a const assertion (after converting to an array instead of a set:

```ts
const VALID_SEARCH_TYPES = ['code', 'repositories'] as const;
```

File: package.json
Comment (jugaltheshah) [reply package.json:49]: It's a bit of a confusing situation. Only `joi` transitioned back out of the `hapi` umbrella. The rest of the modules, including `hoek`, stayed under the Hapi umbrella. You can still find it on the official site [here](https://hapi.dev/module/hoek), and I'll link to the official announcement if I can find it.

**Review by nwalters512 — COMMENTED** [review]
Per https://github.com/NerdWalletOSS/shepherd/issues/352#issuecomment-754351350, leaving a soft review - this looks great! Left a few stylistic comments and one about using `lodash` instead of `@hapi/hoek`, but this is all pretty much exactly what we need. Will hold off on signing off until you let us know it's ready to go.

File: src/adapters/github.ts
Comment (jugaltheshah) [reply src/adapters/github.ts:98]: Sorry I didn't realize `lodash` was already a dependency, will switch!

File: package.json
Comment (aorinevo) [reply package.json:49]: Ah, I see.  

I'll work on migrating `@hapi/joi` back `joi`.

File: src/adapters/github.ts
Comment (jugaltheshah) [reply src/adapters/github.ts:22]: I prefer the suggested way as well, just thought I'd try to be `class-y` :smile:.

File: src/adapters/github.ts
Comment (jugaltheshah) [reply src/adapters/github.ts]: This is complete, resolving.

File: src/adapters/github.ts
Comment (jugaltheshah) [reply src/adapters/github.ts]: This is complete, resolving.

**Comment by jugaltheshah** [comment]
@nwalters512 hello there, thanks for the review & comments. I made the suggested changes and I think the PR's looking pretty good. Only thing I'm not sure of is, what's the expectation for unit test coverage, should every PR come with some?

**Comment by aorinevo** [comment]
> @nwalters512 hello there, thanks for the review & comments. I made the suggested changes and I think the PR's looking pretty good. Only thing I'm not sure of is, what's the expectation for unit test coverage, should every PR come with some?

We would need unit tests covering this new functionality before merging this PR down.

File: package.json
Comment (aorinevo) [inline package.json:78]: Should maintain as much parity as possible between `jest` and `@types/jest`.  Same for other packages with types.

Comment (jugaltheshah) [reply package.json:78]: Updated package typings to latest available. Resolving.

File: src/adapters/git.ts
Comment (jugaltheshah) [inline src/adapters/git.ts:73]: I believe this is necessary because when I update `simple-git` from `v2.13.2` to `v2.31.0`, I get the following error on build:

```bash
src/adapters/git.ts:77:49 - error TS2345: Argument of type '{ '--force': boolean; }' is not assignable to parameter of type 'string[] | Record<string, string | number | null> | undefined'.
  Type '{ '--force': boolean; }' is not assignable to type 'Record<string, string | number | null>'.
    Property ''--force'' is incompatible with index signature.
      Type 'boolean' is not assignable to type 'string | number | null'.

     await this.git(repo).push('origin', 'HEAD', options);
```

I believe this originally stems from slightly incorrect usage of `simple-git` (or that package made a breaking change in a minor update). According to https://github.com/steveukx/git-js#options-as-an-object and the error above, options can be a string array, object or undefined. When an object is passed, key-value pairs should be `string: { string, number, null }`, _not_ `string: boolean`.

The change I propose here should be backwards-compatible while also meeting the expected usage of `simple-git`, but please let me know if not, or whether this should be broken out into its own PR.

Comment (jugaltheshah) [reply src/adapters/git.ts:73]: P.S., was thinking of `const options = !!force ? ['--force'] : undefined;` originally but it seemed redundant as `force` is a boolean. Since Typescript only kicks in at compile time though, is that type coercion advisable?

**Comment by jugaltheshah** [comment]
Added tests. Coverage was not at 100% before this PR, but I believe I've effectively covered this new code.

Here's the result of `npm test` from the last merge to `master`:
```bash
$ npm run test

> @nerdwallet/shepherd@1.11.0 test /home/travis/build/NerdWalletOSS/shepherd

> jest --coverage src/

 PASS  src/util/persisted-data.test.ts

 PASS  src/adapters/github.test.ts

 PASS  src/util/migration-spec.test.ts

 PASS  src/util/generate-pr-message.test.ts

 PASS  src/commands/version.spec.ts

File               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 

-------------------|---------|----------|---------|---------|-------------------

All files          |   39.26 |    21.65 |    37.1 |   39.68 |                   

 adapters          |   20.99 |    14.47 |   17.14 |   22.08 |                   

  git.ts           |   16.67 |        0 |    12.5 |   16.67 | 33-103            

  github.ts        |   21.97 |    15.71 |   18.52 |   23.39 | ...14-155,199-345 

 commands          |     100 |      100 |     100 |     100 |                   

  version.ts       |     100 |      100 |     100 |     100 |                   

 util              |   64.36 |    47.62 |      60 |   65.22 |                   

  ...pr-message.ts |      60 |      100 |      75 |      50 | 10-14             

  ...ation-spec.ts |   67.86 |       50 |      75 |   67.86 | 28-35,47-51       

  octokit.ts       |   14.81 |        0 |       0 |   17.39 | 12-40,47          

  ...isted-data.ts |     100 |      100 |     100 |     100 |                   

-------------------|---------|----------|---------|---------|-------------------

Test Suites: 5 passed, 5 total

Tests:       28 passed, 28 total

Snapshots:   0 total

Time:        6.215 s

Ran all test suites matching /src\//i.

The command "npm run test" exited with 0.
```

And here are the results of `npm test` from this PR:
```bash
$ npm run test

> @nerdwallet/shepherd@1.11.0 test /home/travis/build/NerdWalletOSS/shepherd

> jest --coverage src/

 PASS  src/adapters/github.test.ts

 PASS  src/util/persisted-data.test.ts

 PASS  src/util/migration-spec.test.ts

 PASS  src/util/generate-pr-message.test.ts

 PASS  src/commands/version.spec.ts

File               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 

-------------------|---------|----------|---------|---------|-------------------

All files          |    48.4 |    30.19 |   52.38 |   47.71 |                   

 adapters          |   32.37 |    25.88 |   30.56 |   32.32 |                   

  git.ts           |   16.67 |        0 |    12.5 |   16.67 | 33-101            

  github.ts        |   35.66 |    28.57 |   35.71 |   35.82 | ...34-175,219-365 

 commands          |     100 |      100 |     100 |     100 |                   

  version.ts       |     100 |      100 |     100 |     100 |                   

 util              |   72.28 |    47.62 |      80 |   71.74 |                   

  ...pr-message.ts |      60 |      100 |      75 |      50 | 10-14             

  ...ation-spec.ts |   67.86 |       50 |      75 |   67.86 | 28-35,47-51       

  octokit.ts       |   44.44 |        0 |    62.5 |   43.48 | 12,22-36          

  ...isted-data.ts |     100 |      100 |     100 |     100 |                   

-------------------|---------|----------|---------|---------|-------------------

Test Suites: 5 passed, 5 total

Tests:       31 passed, 31 total

Snapshots:   0 total

Time:        6.647 s

Ran all test suites matching /src\//i.

The command "npm run test" exited with 0.
``` 

Please re-review @aorinevo @nwalters512.

File: src/adapters/git.ts
Comment (aorinevo) [reply src/adapters/git.ts:73]: I'd recommend splitting this out into a separate PR to mitigate risk and keep the focus on the feature.

If you're up for doing so, it might also be good to roll in the change for `joi` (see [1](https://github.com/NerdWalletOSS/shepherd/issues/354)).

References
1. https://github.com/NerdWalletOSS/shepherd/issues/354

Comment (aorinevo) [reply src/adapters/git.ts:73]: > P.S., was thinking of `const options = !!force ? ['--force'] : undefined;` originally but it seemed redundant as `force` is a boolean. Since Typescript only kicks in at compile time though, is that type coercion advisable?

What are alternatives to coercion?

Comment (jugaltheshah) [reply src/adapters/git.ts]: In my mind, coercion is a bit redundant, so the alternatives are to do coercion at all or not.  We could also set a sane default and do input validation for the force parameter.

Comment (jugaltheshah) [reply src/adapters/git.ts]: I reverted the `simple-git` update however, so closing this and will reference in a new PR.

**Comment by aorinevo** [comment]
Tested locally on public repos.  Works as expected.

**Comment by aorinevo** [comment]
Completed code review for this PR.  Looks good to me.  

Many thanks for your contributions @jugaltheshah!

**Review by aorinevo — APPROVED** [review]

**Review by aorinevo — APPROVED** [review]

**Comment by aorinevo** [comment]
@nwalters512, mind if I merge this down?

**Comment by aorinevo** [comment]
Going ahead and merging down to upstream.

**Pull request merged**
